### DISCUSSION TITLE
back_population of secondary relationships conflicts

### DISCUSSION THREAD

**Discussion by MaicoTimmerman** [sqlalchemy/sqlalchemy#6424] [Usage Questions]
Given this many-to-many setup of `A -> C <- B`:
```py
from sqlalchemy import Column, create_engine, ForeignKey, select
from sqlalchemy import Integer
from sqlalchemy.orm import registry, Session, relationship

mapper_registry: registry = registry()

e = create_engine('sqlite:///database_relationship_secondary.db')

@mapper_registry.mapped
class A:
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)
    all_b = relationship("B", secondary="c")


@mapper_registry.mapped
class B:
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)

@mapper_registry.mapped
class C:
    __tablename__ = "c"
    a_id = Column(Integer, ForeignKey("a.id"), primary_key=True)
    b_id = Column(Integer, ForeignKey("b.id"), primary_key=True)
    a = relationship("A")
    b = relationship("B")

mapper_registry.metadata.create_all(e)
#
with Session(e) as s:
    s.execute(select(C)).all()

```

I'm getting two warnings for errors of conflicting relationships:
```
SAWarning: relationship 'C.a' will copy column a.id to column c.a_id, which conflicts with relationship(s): 'A.all_b' (copies a.id to c.a_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   The 'overlaps' parameter may be used to remove this warning. (Background on this error at: http://sqlalche.me/e/14/qzyx)
SAWarning: relationship 'C.b' will copy column b.id to column c.b_id, which conflicts with relationship(s): 'A.all_b' (copies b.id to c.b_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   The 'overlaps' parameter may be used to remove this warning. (Background on this error at: http://sqlalche.me/e/14/qzyx)
```

[The link](https://docs.sqlalchemy.org/en/14/errors.html#relationship-x-will-copy-column-q-to-column-p-which-conflicts-with-relationship-s-y) only mentions overlapping cases, not the secondary case. What would be the correct way of adding `back_populates`, or should some of these become `viewonly`?

**Comment by zzzeek** [comment]
this just came up also in #6400 and we should use that to continue work on improving the messaging here.   the warning was intended to be a warning for relationships that may have intended to use back_populates, but then it also ended up covering the case where someone uses "secondary" at the same time as relationships directly on the class that's mapped directly to "secondary".    that wasn't what I was going for, but like so many other cases i was being short sighted, because this is definitely something that is worth warning for, so we definitely need to add this to the docs, and to the warning message, and to the tests.        I think usually you'd want to put "viewonly" on the "secondary" relationship, since otherwise if you create new C objects you are conflicting with mutations made to A.all_b directly.

also note the warning only comes up if C actually has relationships on it, even though if it didn't, the existence of C as a mapping still introduces the possibility of conflicting rows.   however, that warning would be not as easy to hide because you can't use "overlaps" to silence it, so im OK with that case not warning right now.

**Reply by zzzeek** [reply]
someone also made an example case at #6430.
